ibtk/src/utilities/RestartCleaner.cpp

```diff
@@ -0,0 +1,269 @@
+    : d_object_name(object_name),
+      d_restart_base_path(),
+      d_strategy(parseStrategy(input_db->getStringWithDefault("cleanup_strategy", "KEEP_RECENT_N"))),
+      d_keep_restart_count(input_db->getIntegerWithDefault("keep_recent_files", 5)),
+      d_enabled(input_db->getBoolWithDefault("enable_cleaner", false)),
+      d_log_actions(input_db->getBoolWithDefault("log_cleaning_actions", true)),
+      d_dry_run(input_db->getBoolWithDefault("dry_run", false))
+{
+    if (d_enabled)
+    {
+        if (input_db->keyExists("restart_directory"))
+        {
+            d_restart_base_path = input_db->getString("restart_directory");
+        }
+        else
+        {
+            TBOX_ERROR(d_object_name << "::RestartCleaner(): "
+                                     << "'restart_directory' must be specified when enable_cleaner is true"
+                                     << std::endl);
+        }
+    }
+} // RestartCleaner
+
+void
+RestartCleaner::cleanup()
+{
+    if (!d_enabled) return;
+
+    // Only perform file system operations on the master processor
+    if (IBTK_MPI::getRank() == 0)
+    {
+        executeStrategy();
+    }
+    IBTK_MPI::barrier();
+} // cleanup
+
+std::vector<int>
+RestartCleaner::getAvailableIterations() const
+{
+    std::vector<int> iterations;
+
+    if (IBTK_MPI::getRank() == 0)
+    {
+        auto dirs = getAllRestartDirs(d_restart_base_path);
+        for (const auto& dir : dirs)
+        {
+            int iter = parseIterationNum(dir.filename().string());
+            if (iter >= 0)
+            {
+                iterations.push_back(iter);
+            }
+        }
+        std::sort(iterations.begin(), iterations.end());
+    }
+
+    int vec_size = static_cast<int>(iterations.size());
+    vec_size = IBTK_MPI::bcast(vec_size, 0);
+
+    // Resize vector on non-master processors
+    if (IBTK_MPI::getRank() != 0)
+    {
+        iterations.resize(vec_size);
+    }
+
+    if (vec_size > 0)
+    {
+        IBTK_MPI::bcast(iterations.data(), vec_size, 0);
+    }
+
+    return iterations;
+} // getAvailableIterations
+
+bool
+RestartCleaner::isEnabled() const
+{
+    return d_enabled;
+} // isEnabled
+
+/////////////////////////////// PRIVATE //////////////////////////////////////
+
+RestartCleaner::CleanupStrategy
+RestartCleaner::parseStrategy(const std::string& strategy_str) const
+{
+    if (strategy_str == "KEEP_RECENT_N")
+    {
+        return CleanupStrategy::KEEP_RECENT_N;
+    }
+
+    TBOX_ERROR(d_object_name << "::parseStrategy(): Unknown strategy: " << strategy_str << std::endl);
+    return CleanupStrategy::KEEP_RECENT_N; // Never reached
+} // parseStrategy
+
+void
+RestartCleaner::executeStrategy() const
+{
+    switch (d_strategy)
+    {
+    case CleanupStrategy::KEEP_RECENT_N:
+        keepRecentN();
+        break;
+    default:
+        TBOX_ERROR(d_object_name << "::executeStrategy(): Unknown strategy" << std::endl);
+    }
+} // executeStrategy
+
+int
```

Comment:
```suggestion
std::optional<int>
```
Let's encode a missing directory number with `std::optional`: that way we don't have to use `-1` as a magic value.